src/lib/services/client-metrics/index.ts

```diff
@@ -1,55 +1,117 @@
 
-'use strict';
-
-const Projection = require('./projection');
 const TTLList = require('./ttl-list');
-const appSchema = require('./metrics-schema');
-const { clientMetricsSchema } = require('./client-metrics-schema');
 const { clientRegisterSchema } = require('./register-schema');
-const { APPLICATION_CREATED } = require('../../types/events');
 
 const FIVE_SECONDS = 5 * 1000;
 const FIVE_MINUTES = 5 * 60 * 1000;
 
-module.exports = class ClientMetricsService {
+export interface IAppInstance {
+    appName: string;
+    instanceId: string;
+    sdkVersion: string;
+    clientIp: string;
+    lastSeen: Date;
+    createdAt: Date;
+}
+
+export interface IClientApp {
+    appName: string;
+    instanceId: string;
+    clientIp?: string;
+    seenToggles?: string[];
+    metricsCount?: number;
+    strategies?: string[] | Record<string, string>[];
+    bucket?: any;
+    count?: number;
+    started?: number | Date;
+    interval?: number;
+    icon?: string;
+    description?: string;
+    color?: string;
+}
+
+export interface IAppFeature {
+    name: string;
+    description: string;
+    type: string;
+    project: string;
+    enabled: boolean;
+    stale: boolean;
+    strategies: any;
+    variants: any[];
+    createdAt: Date;
+    lastSeenAt: Date;
+}
+
+export interface IApplicationQuery {
+    strategyName?: string;
+}
+
+export interface IAppName {
+    appName: string;
+}
+
+export interface IMetricCounts {
+    yes?: number;
+    no?: number;
+    variants?: Record<string, number>;
+}
+
+export interface IMetricsBucket {
+    start: Date;
+    stop: Date;
+    toggles: IMetricCounts;
+}
+
+export default class ClientMetricsService {
+    globalCount = 0;
+
+    apps = {};
+
+    lastHourProjection = new Projection();
+
+    lastMinuteProjection = new Projection();
+
+    lastHourList = new TTLList({
+        interval: 10000,
+    });
+
+    logger = null;
+
+    lastMinuteList = new TTLList({
+        interval: 10000,
+        expireType: 'minutes',
+        expireAmount: 1,
+    });
+
+    seenClients: Record<string, IClientApp> = {};
+
+    private timers: NodeJS.Timeout[] = [];
+
     constructor(
-        {
-            clientMetricsStore,
-            strategyStore,
-            featureToggleStore,
-            clientApplicationsStore,
-            clientInstanceStore,
-            eventStore,
-        },
-        {
-            getLogger,
-            bulkInterval = FIVE_SECONDS,
-            announcementInterval: appAnnouncementInterval = FIVE_MINUTES,
-        },
+        private clientMetricsStore: ClientMetricsStore,
+        private strategyStore: StrategyStore,
+        private toggleStore: FeatureToggleStore,
+        private clientAppStore: ClientApplicationsDb,
+        private clientInstanceStore: ClientInstanceStore,
+        private eventStore: EventStore,
+        private getLogger: LogProvider,
+        private bulkInterval = FIVE_SECONDS,
+        private announcementInterval = FIVE_MINUTES,
```

Comment:
I would have prefered the constructor to be in line with the format utilized by all the other services converted to typescript, something in the line of:

```
---
constructor(
        stores: Pick<IUnleashStores,
           | 'clientMetricsStore'
          | 'strategyStore'
          | 'featureToggleStore'
          | 'clientApplicationsStore'
          | 'clientInstanceStore'
          | 'eventStore'
        >,
        config: Pick<IUnleashConfig, 'getLogger'>,
    ) {
        this.clientMetricsStore = stores.clientMetricsStore;
        this.strategyStore = stores.strategyStore;
        this.toggleStore = stores.featureToggleStore;
        this.clientAppStore = stores.clientApplicationsStore;
        this.clientInstanceStore = stores.clientInstanceStore;
        this.eventStore = stores.eventStore;
----
```

Reply:
Your proposal mirrors the JS approach, but I don't think it makes the class easier to use, especially since the `stores` variable is immediately unwrapped into the individual stores.

My change just moved the unwrapping to outside the constructor.

I'm fine if you want it changed back for consistency, but I think simplifying the classes for clarity and testing is worth doing.

Reply:
With the Pick format it will actually be the same from the test perspective. You will only have to provide the requirements from the Picks and you will get the type-support from TypeScript.

Reply:
Interesting! So in the tests I won't have to create an `IUnleashStores` with all keys before the pick? Or would you use and `any` type on that?

Learning more here: https://www.typescriptlang.org/docs/handbook/utility-types.html#picktype-keys

Since you are in the code, are you going to make that change?

Reply:
> Since you are in the code, are you going to make that change?

On it!

Reply:
You can see the result here, basically the test provides the same as before, just wrappted in objects. 

https://github.com/Unleash/unleash/pull/831/commits/644cd27a395c2d635f32e2fe793a5e07a3c1b902